### PR TITLE
Input

### LINKED ISSUE #1: There’s no mechanism for user input
It’s just a tree thus far, not actually anything describing a UI.

### PR DESCRIPTION
**Pull request by robrix**
Adds a trivial model for taking textual input.
- [x] Fixes #1.

### CODE REVIEW

**Comment by robrix** [comment]
cc @andymatuschak

File: app/Main.hs [hunk 1]
```diff
@@ -1,32 +1,26 @@
...
 
-toString :: Show a => View a -> String
-toString = intercalate "\n" . iter go . fmap (pure . show)
-  where go (Text s) = [ s ]
-        go (List vs) = vs >>= fmap ("- " ++)
-
-toHTML :: Show a => View a -> String
-toHTML = intercalate "\n" . iter go . fmap (pure . show)
-  where go (Text s) = [ "<p>" ++ s ++ "</p>" ]
-        go (List vs) = "<ul>" : (vs >>= fmap (\ s -> "  <li>" ++ s ++ "</li>")) ++ ["</ul>"]
```
Comment (robrix) [inline app/Main.hs:32]: I didn’t feel like spinning up a server to try out the request/response flow implied by `Input`. Might try again later on down the line with `Snap`.

File: app/Main.hs [hunk 2]
```diff
@@ -1,32 +1,27 @@
...
+          $   flag CLI CLI (long "cli" <> short 'C')))
           (fullDesc <> progDesc "UI playground")
 
 run :: Arguments -> IO ()
-run Arguments{..} = putStrLn $ case format of
-  String -> toString (list [ text "hello", text "world" ] :: View ())
-  HTML -> toHTML (list [ text "hello", text "world" ] :: View ())
+run Arguments{..} = case format of
+  CLI -> cli view
+  where view = list [ text "hello", input >>= text ] :: View ()
```
Comment (robrix) [inline app/Main.hs:21]: It prompts:

```
$ stack exec ui-effects
hello
> world
world
```

**Comment by robrix** [comment]
It’ll be interesting to see how this feels when you hit up against the expression problem.

**Comment by andymatuschak** [comment]
> It’ll be interesting to see how this feels when you hit up against the expression problem.

Yes, seems like the definition in src/UI/View.hs requires a closed set of node types.

`Control.Monad.Free.Church` was new to me—thank you for that!

**Pull request merged**